### PR TITLE
Local build

### PR DESCRIPTION
**Pull request by p-netm**
Goal :
to have the react app built on the controlling machine, and then have the built sent to the target host.

### CODE REVIEW

**Comment by p-netm** [comment]
one problem am currently facing is that:
- for the build to happen on the controlling machine , the role need to setup node and yarn on the controlling machine.
- I am currently trying to see how that plays out with `molecule converge` for a remote build using nvm. 
- the problem it seems ; am not able to get nvm to play nice: specifically the task that specifically installs nvm.
- I am having trouble with the install.yml file; the pending part is to make sure the required pre-requisite packages(*node*, yarn, git) are setup on the either the local machine or the controlled machine as needed.

Screenshot from 2020-08-19 13-12-39.

**Comment by moshthepitt** [comment]
It seems that `/home/react/.nvm/nvm.sh` isn't found.  Have you set `nvm_install_path` okay?

**Comment by p-netm** [comment]
> It seems that `/home/react/.nvm/nvm.sh` isn't found. Have you set `nvm_install_path` okay?

`/home/react/.nvm/nvm.sh` is dynamically created in the install nvm task

**Comment by moshthepitt** [comment]
@p-netm what is causing the failing test?

**Comment by p-netm** [comment]
> @p-netm what is causing the failing test?

@moshthepitt 
- either the tag in the test/test.yml file was not correct,
- when `yarn build` required some envs to be set for e.g `SKIP_PREFLIGHT_CHECK`
- later tags of kaznet would complain with some SASS transpilation error

- I fixed the tag to kaznet-frontend `v0.01`, did not seem like it had the above errors.

**Comment by moshthepitt** [comment]
@p-netm any reason why you prefer using Kaznet as opposed to Reveal in the test?

**Comment by p-netm** [comment]
> @p-netm any reason why you prefer using Kaznet as opposed to Reveal in the test?

nothing important really, kaznet build seems to be lighter and it was the initial value

**Review by moshthepitt — APPROVED** [review]
Looks good to me but lets get a review from @ukanga and/or @morrismukiri

File: defaults/main.yml [hunk 1]
```diff
@@ -1,39 +1,37 @@
...
-react_nvm_node_dir: "{{ react_nvm_nvm_root }}/versions/node/{{ react_nvm_node_version }}"
-react_nvm_npm_path: "{{ react_node_dir }}/bin/npm"
-react_nvm_version: "v0.34.0"
+react_local_user_home: "{{ lookup('env', 'HOME') }}"
+react_local_user: "{{ lookup('env','USER') }}"
 
 # git
 react_git_url:
-react_git_version: "master"
+react_git_version: "v0.6.2-rc1"
```
Comment (moshthepitt) [inline defaults/main.yml:10]: Should we keep this as master

File: defaults/main.yml [hunk 2]
```diff
@@ -1,39 +1,37 @@
...
 react_app_name: "{{ react_system_user }}"
 react_codebase_path: "{{ react_system_user_home }}/app"
 react_versioned_path: "{{ react_codebase_path }}-versioned"
 react_checkout_path: "{{ react_versioned_path }}/{{ ansible_date_time['epoch'] }}"
 react_log_path: "/var/log/{{ react_app_name }}"
+react_max_versioned_folders: 10
 
 # app settings
 react_app_settings:
-  REACT_APP_WEBSITE_NAME: 'React Example App'
```
Comment (moshthepitt) [inline defaults/main.yml:39]: Any reason why this is being removed?

File: molecule/default/molecule.yml [hunk 3]
```diff
@@ -3,19 +3,22 @@ dependency:
   name: galaxy
 driver:
   name: docker
-lint:
-  name: yamllint
 platforms:
   - name: ubuntu-16.04
```
Comment (moshthepitt) [inline molecule/default/molecule.yml:7]: Should we change the name?

File: tasks/cleanup.yml [hunk 4]
```diff
@@ -2,11 +2,15 @@
 - name: Get previous app deploy listing
   command: ls -tr1 chdir={{ react_versioned_path }}
   register: versioned_list
-  changed_when: "versioned_list.stdout_lines|length > 1"
+  changed_when: false
```
Comment (moshthepitt) [inline tasks/cleanup.yml:5]: Why this change?

File: tasks/configure.yml [hunk 5]
```diff
@@ -1,24 +1,78 @@
 ---
-- name: Copy environment variables file
```
Comment (moshthepitt) [inline tasks/configure.yml:2]: If we remove this, will things still work for non-local builds?

File: tasks/cleanup.yml (hunk 4 above)
Comment (p-netm) [reply tasks/cleanup.yml:5]: This task is setting a fact, it will always report as changed on any run.

File: tasks/configure.yml (hunk 5 above)
Comment (p-netm) [reply tasks/configure.yml:2]: yeah I suppose so.

Comment (p-netm) [reply tasks/configure.yml:2]: why?

File: tasks/cleanup.yml (hunk 4 above)
Comment (moshthepitt) [reply tasks/cleanup.yml:5]: cool

File: tasks/configure.yml (hunk 5 above)
Comment (moshthepitt) [reply tasks/configure.yml:2]: Just wondering because I dont want to break non local builds

File: defaults/main.yml [hunk 6]
```diff
@@ -22,18 +13,26 @@ react_git_key_filename: "id_ed25519"
 react_remove_git_key: true
 
 # system-wide dependencies
-react_system_wide_dependencies:
+react_remote_system_wide_dependencies:
   - yarn
```
Comment (ukanga) [inline defaults/main.yml:17]: Do you not require yarn to be installed locally when building locally?

File: defaults/main.yml [hunk 7]
```diff
@@ -22,18 +13,26 @@ react_git_key_filename: "id_ed25519"
...
+  - git
+react_local_system_wide_dependencies:
+  - git
 react_node_version: 10.x
 
 # app
+react_remote_js_build: True
+react_local_checkout_path: "{{ react_local_user_home }}/reveal"
```
Comment (ukanga) [inline defaults/main.yml:25]: Could we have the local checkout in a temporary directory that potentially gets removed after the build process? Specifically in the `/tmp/$USER/reveal` directory?
A git clone in `/tmp` folder is the approach taken in the ansible-superset role, trying to understand why the home folder is an option here.
Or perhaps, `~/.cache/reveal` folder if caching is the intention.

Comment (ukanga) [reply defaults/main.yml:25]: I prefer `/tmp` over `.cache`.

Comment (p-netm) [reply defaults/main.yml:25]: cool.

File: tasks/configure.yml (hunk 5 above)
Comment (p-netm) [reply tasks/configure.yml:2]: non local builds should still run fine

File: defaults/main.yml (hunk 6 above)
Comment (p-netm) [reply defaults/main.yml:17]: when building locally, node and yarn are a pre-requisite, that should be set up before hand.
I did not have any good idea on how to include tasks to set them up in this role(i.e. for local builds).
for admin host, I have installed yarn as a global npm package after having installed node using nvm.
- problem is that ansible threw an error when I tried to add an include_role task that installs nvm and that is delegated to localhost.

Comment (ukanga) [reply defaults/main.yml:17]: The wording looks okay but I suppose it requires an actual test on my end to know for sure that there is absolutely a need to stick to the previous variable.

File: defaults/main.yml [hunk 8]
```diff
@@ -22,18 +13,26 @@ react_git_key_filename: "id_ed25519"
...
+react_remote_system_wide_dependencies:
   - yarn
+  - git
+react_local_system_wide_dependencies:
+  - git
 react_node_version: 10.x
 
 # app
+react_remote_js_build: True
+react_local_checkout_path: "/tmp/reveal"
```
Comment (moshthepitt) [inline defaults/main.yml:25]: Is it possible to make this something like `/tmp/$USER/{ the react app name }`?

Comment (p-netm) [reply defaults/main.yml:25]: yeah sure.
